src/ripple/basics/impl/UptimeTimer.cpp

```diff
@@ -19,64 +19,53 @@
 
 #include <ripple/basics/UptimeTimer.h>
 
-#include <atomic>
-
 namespace ripple {
 
-UptimeTimer::UptimeTimer ()
-    : m_elapsedTime (0)
-    , m_startTime (::time (0))
-    , m_isUpdatingManually (false)
-{
-}
+std::atomic<UptimeClock::rep>  UptimeClock::now_{0};      // seconds since start
+std::atomic<bool>              UptimeClock::stop_{false}; // stop update thread
 
-UptimeTimer::~UptimeTimer ()
+// On rippled shutdown, cancel and wait for the update thread
+UptimeClock::update_thread::~update_thread()
 {
-}
-
-int UptimeTimer::getElapsedSeconds () const
-{
-    int result;
-
-    if (m_isUpdatingManually)
-    {
-        std::atomic_thread_fence (std::memory_order_seq_cst);
-        result = m_elapsedTime;
-    }
-    else
+    if (joinable())
     {
-        // VFALCO TODO use time_t instead of int return
-        result = static_cast <int> (::time (0) - m_startTime);
+        stop_ = true;
+        join();
     }
-
-    return result;
-}
-
-void UptimeTimer::beginManualUpdates ()
-{
-    //assert (!m_isUpdatingManually);
-
-    m_isUpdatingManually = true;
 }
 
-void UptimeTimer::endManualUpdates ()
+// Launch the update thread
+UptimeClock::update_thread
+UptimeClock::start_clock()
 {
-    //assert (m_isUpdatingManually);
-
-    m_isUpdatingManually = false;
+    return update_thread{[]
+                         {
+                             using namespace std;
+                             using namespace std::chrono;
+
+                             // Wake up every second and update now_
+                             auto next = system_clock::now() + 1s;
```

Comment:
I chose `system_clock` because I want to take advantage of clock corrections over days, weeks and months.  This will keep the uptime accurate to within 1s.  Use of `steady_clock` is prone to drifting because all clocks are inaccurate, but `steady_clock` can't be corrected.  So over weeks and months the uptime would get off by a few seconds.  I note that my quartz watch drifts by a second or two per week (if not corrected), and I'm figuring `steady_clock` is no more accurate than that.

Reply:
🙇 :+1: